src/OauthTokenStorage.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 2 as published by the
+ * Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_edge;
+
+use Apigee\Edge\Exception\OauthAuthenticationException;
+use Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface;
+use Drupal\apigee_edge\Plugin\EdgeOauthTokenKeyTypeInterface;
+use Drupal\Core\Entity\EntityStorageException;
+use Drupal\key\KeyInterface;
+
+/**
+ * Storing and returning OAuth access token data.
+ */
+class OauthTokenStorage implements OauthTokenStorageInterface {
+
+  /**
+   * The timestamp when the token expires.
+   *
+   * This value is calculated from "expires_in" value which is the lifetime of the access token in seconds.
+   *
+   * @var int
+   */
+  protected $expires = 0;
+
+  /**
+   * Number of seconds extracted from token's expiration date when hasExpired() calculates.
+   *
+   * This ensures that token gets refreshed earlier than it expires.
+   *
+   * @var int
+   */
+  protected $leeway;
+
+  /**
+   * The OAuth token key entity.
+   *
+   * @var \Drupal\key\KeyInterface
+   */
+  protected $key;
+
+  /**
+   * The key type of the OAuth token key entity.
+   *
+   * @var \Drupal\apigee_edge\Plugin\EdgeOauthTokenKeyTypeInterface
+   */
+  protected $keyType;
+
+  /**
+   * OauthTokenStorage constructor.
+   *
+   * @param \Drupal\key\KeyInterface $key
+   *   The OAuth token key entity.
+   * @param int $leeway
+   *   Number of seconds extracted from token's expiration date.
+   */
+  public function __construct(KeyInterface $key, int $leeway = 30) {
+    if (!($key->getKeyType() instanceof EdgeOauthTokenKeyTypeInterface)) {
+      throw new \InvalidArgumentException("Type of {$key->id()} key does not implement EdgeOauthTokenKeyTypeInterface.");
+    }
+
+    $this->key = $key;
+    $this->keyType = $key->getKeyType();
+    $this->leeway = $leeway;
```

Comment:
Because the value of the leeway could not come elsewhere it should come from the OAuth key storage I think. 
(It is not necessary to be fixed now just leave a TODO comment here, remove it from the constructor and set its default value to 30s.)